calyx/src/passes/parseq_to_seqpar.rs

```diff
@@ -0,0 +1,116 @@
+use crate::ir;
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+use crate::ir::GetAttributes;
+
+#[derive(Default)]
+/// Transforms a par of seq blocks into a seq of par blocks. For the
+/// transformation to occur, each seq block must have the same number of statements,
+/// and the nth statement of each seq block in the par block
+/// must all have the same number of clock cycles.
+///
+/// #Example
+/// par {
+///    seq { @static(M) A0; @static(N) B0; @static(P) C0; }
+///    seq { @static(M) A1; @static(N) B1; @static(P) C1; }
+///    seq { @static(M) A2; @static(N) B2; @static(P) C2; }
+///}
+///
+/// into
+///
+/// seq {
+///     par{ @static(M) A0; @static(M) A1; @static(M) A2;}
+///     par{ @static(N) B0; @static(N) B1; @static(N) B2;}
+///     par{ @static(P) C0; @static(P) C1; @static(P) C2;}
+/// }
+///  
+pub struct ParSeqToSeqPar;
+
+impl Named for ParSeqToSeqPar {
+    fn name() -> &'static str {
+        "parseq-to-seqpar"
+    }
+
+    fn description() -> &'static str {
+        "Transform `par` of `seq` to `seq` of `par` under correct conditions"
+    }
+}
+
+impl Visitor for ParSeqToSeqPar {
+    fn finish_par(
+        &mut self,
+        s: &mut ir::Par,
+        _comp: &mut ir::Component,
+        _sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let mut new_seq = ir::Seq {
+            stmts: Vec::new(),
+            attributes: ir::Attributes::default(),
+        };
+
+        let mut transformation_ok = true;
+        let mut all_seqs_iterated = false;
+        let mut n = 0;
+
+        while !all_seqs_iterated && transformation_ok {
```

Comment:
If we separate the code into two phases like I described below, the transformation part of the code should iterate over a vector of vectors of bodies of seq statements instead of using this `all_seqs_iterated` stuff.